### PR TITLE
workflow for enrichment analysis using goseq

### PR DESCRIPTION
**Pull request by nilchia**
This workflow uses goseq to do GO and KEGG enrichment analysis on mm10 genome.

### CODE REVIEW

File: workflows/transcriptomics/goseq/README.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+            ```
+ 
+## Processing
+
+- The workflow will do a simple enrichment analysis with taking into account the gene length
+- The output will be 3 files `GO table`, `Top ontology plot` and `DE genes in each category` for CC, BP, and MF ontologies and `KEGG table` and `DE genes in each KEGG Pathways`
+
+### Warning
+
+- This workflow uses mm10 reference genome
```
Comment (mvdbeek) [inline workflows/transcriptomics/goseq/README.md:25]: Can you please make that configurable ? You can insert a simple workflow parameter and select "Attempt restriction base on connections"

File: workflows/transcriptomics/goseq/plnmotmptestjobwtl761pz.json
Comment (mvdbeek) [inline workflows/transcriptomics/goseq/plnmotmptestjobwtl761pz.json:1]: You can delete this

File: workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis-tests.yml [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            - that: "has_text"
+              text: "category	over_represented_pvalue	under_represented_pvalue	numDEInCat	numInCat	term	ontology	p_adjust_over_represented	p_adjust_under_represented"
+            - that: "has_text"
+              text: "GO:0005737	4.7986980823087e-08	1	4810	8312	cytoplasm	CC	9.31427297776118e-05	1"
+            - that: "has_text"
+              text: "over_represented_pvalue"
+            - that: "has_text"
+              text: "under_represented_pvalue"
+    goseq-plot-cc:
+      path: test-data/goseq-plot-cc.pdf
```
Comment (mvdbeek) [inline workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis-tests.yml:29]: Can you use a size assertions here and delete the PDF in  the test-data folder?

File: workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis-tests.yml [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    gene length:
+      class: File
+      path: test-data/gene-length.tabular
+      filetype: tabular
+    KEGG pathways:
+      class: File
+      path: test-data/kegg-pathways.tabular
+      filetype: tabular
+  outputs:
+    goseq-table-cc:
```
Comment (mvdbeek) [inline workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis-tests.yml:16]: These are the primary outputs, can you adjust them (here and in the workflow) to be human readable, i.e. no `-`, `_` and abbreviations that aren't well known ? I don't know what CC is here

File: workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis-tests.yml [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    goseq-cat-genes-kegg:
+      element_tests:
+        goseq-cat-genes-kegg.tabular:
+          asserts:
+            - that: "has_text"
+              text: "category	de_genes"
+            - that: "has_text"
+              text: "04540	ENSMUSG00000063358,ENSMUSG00000045136,ENSMUSG00000025856,ENSMUSG00000028019"
+    goseq-plot-mf:
+      path: test-data/goseq-plot-mf.pdf
```
Comment (mvdbeek) [inline workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis-tests.yml:69]: Same here, just a size assertion should be fine

File: workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis-tests.yml [hunk 5]
```diff
@@ -0,0 +1,101 @@
+- doc: Test outline for goseq-go-kegg-enrichment-analsis
+  job:
+    DE result:
```
Comment (mvdbeek) [inline workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis-tests.yml:3]:
```suggestion
    Differential expression result:
```

File: workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis-tests.yml [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+      path: test-data/de-result.tabular
+      filetype: tabular
+    gene length:
+      class: File
+      path: test-data/gene-length.tabular
+      filetype: tabular
+    KEGG pathways:
+      class: File
+      path: test-data/kegg-pathways.tabular
+      filetype: tabular
```
Comment (mvdbeek) [inline workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis-tests.yml:14]: Can you upload these to zenodo and use the link as input ?

**Review by mvdbeek — COMMENTED** [review]
Looking good, thank you!

File: workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis.ga [hunk 7]
```diff
@@ -0,0 +1,568 @@
...
+    "format-version": "0.1",
+    "license": "MIT",
+    "release": "0.1",
+    "name": "goseq-go-kegg-enrichment-analsis",
+    "report": {
+        "markdown": "\n# Workflow Execution Report\n\n## Workflow Inputs\n```galaxy\ninvocation_inputs()\n```\n\n## Workflow Outputs\n```galaxy\ninvocation_outputs()\n```\n\n## Workflow\n```galaxy\nworkflow_display()\n```\n"
+    },
+    "steps": {
+        "0": {
+            "annotation": "input1: table with two columns. 1 gene 2 T/F based on DEA (is it DEG?)",
```
Comment (lldelisle) [inline workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis.ga:21]:
```suggestion
            "annotation": "input1: table with two columns. 1 gene 2 T/F based on Differential Expression Analysis",
```

File: workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis.ga [hunk 8]
```diff
@@ -0,0 +1,568 @@
...
+            "content_id": null,
+            "errors": null,
+            "id": 0,
+            "input_connections": {},
+            "inputs": [
+                {
+                    "description": "input1: table with two columns. 1 gene 2 T/F based on DEA (is it DEG?)",
```
Comment (lldelisle) [inline workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis.ga:28]:
```suggestion
                    "description": "input1: table with two columns. 1 gene 2 T/F based on Differential Expression Analysis",
```

File: workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis.ga [hunk 9]
```diff
@@ -0,0 +1,568 @@
...
+            "tool_id": null,
+            "tool_state": "{\"optional\": false, \"format\": [\"tabular\"], \"tag\": null}",
+            "tool_version": null,
+            "type": "data_input",
+            "uuid": "ce8469f0-065c-4c4d-a186-bea4ed476868",
+            "when": null,
+            "workflow_outputs": []
+        },
+        "3": {
+            "annotation": "input1: table with two columns. 1 gene 2 T/F based on DEA (is it DEG?)\n\ninput2: table with two columns. 1 gene 2 gene length",
```
Comment (lldelisle) [inline workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis.ga:102]:
```suggestion
            "annotation": "input1: table with two columns. 1 gene 2 T/F based on DEA\n\ninput2: table with two columns. 1 gene 2 gene length",
```

File: workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis.ga [hunk 10]
```diff
@@ -0,0 +1,568 @@
...
+                    "id": 0,
+                    "output_name": "output"
+                },
+                "length_file": {
+                    "id": 1,
+                    "output_name": "output"
+                }
+            },
+            "inputs": [],
+            "label": "goseq - CC",
```
Comment (lldelisle) [inline workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis.ga:117]:
```suggestion
            "label": "goseq - Cellular Component",
```

File: workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis.ga [hunk 11]
```diff
@@ -0,0 +1,568 @@
...
+                },
+                {
+                    "label": "goseq-cat-genes-cc",
+                    "output_name": "cat_genes_tab",
+                    "uuid": "bff6fd1b-d67b-4ca1-a108-853e0ab61377"
+                }
+            ]
+        },
+        "4": {
+            "annotation": "input1: table with two columns. 1 gene 2 T/F based on DEA (is it DEG?)\n\ninput2: table with two columns. 1 gene 2 gene length",
```
Comment (lldelisle) [inline workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis.ga:212]:
```suggestion
            "annotation": "input1: table with two columns. 1 gene 2 T/F based on DEA\n\ninput2: table with two columns. 1 gene 2 gene length",
```

File: workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis.ga [hunk 12]
```diff
@@ -0,0 +1,568 @@
...
+                    "id": 0,
+                    "output_name": "output"
+                },
+                "length_file": {
+                    "id": 1,
+                    "output_name": "output"
+                }
+            },
+            "inputs": [],
+            "label": "goseq - BP",
```
Comment (lldelisle) [inline workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis.ga:227]:
```suggestion
            "label": "goseq - Biological Process",
```

File: workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis.ga [hunk 13]
```diff
@@ -0,0 +1,568 @@
...
+                },
+                {
+                    "label": "goseq-table-bp",
+                    "output_name": "wallenius_tab",
+                    "uuid": "6922dc30-d93d-4d77-ad26-b7fbd9c6bf13"
+                }
+            ]
+        },
+        "5": {
+            "annotation": "input1: table with two columns. 1 gene 2 T/F based on DEA (is it DEG?)\n\ninput2: table with two columns. 1 gene 2 gene length",
```
Comment (lldelisle) [inline workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis.ga:322]:
```suggestion
            "annotation": "input1: table with two columns. 1 gene 2 T/F based on DEA\n\ninput2: table with two columns. 1 gene 2 gene length",
```

File: workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis.ga [hunk 14]
```diff
@@ -0,0 +1,568 @@
...
+                    "id": 0,
+                    "output_name": "output"
+                },
+                "length_file": {
+                    "id": 1,
+                    "output_name": "output"
+                }
+            },
+            "inputs": [],
+            "label": "goseq - MF",
```
Comment (lldelisle) [inline workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis.ga:337]:
```suggestion
            "label": "goseq - Molecular Function",
```

File: workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis.ga [hunk 15]
```diff
@@ -0,0 +1,568 @@
...
+                },
+                {
+                    "label": "goseq-table-mf",
+                    "output_name": "wallenius_tab",
+                    "uuid": "ced68bff-2dea-4af9-9135-e32f9efa5c7a"
+                }
+            ]
+        },
+        "6": {
+            "annotation": "input1: table with two columns. 1 gene 2 T/F based on DEA (is it DEG?)\n\ninput2: table with two columns. 1 gene 2 gene length",
```
Comment (lldelisle) [inline workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis.ga:432]:
```suggestion
            "annotation": "input1: table with two columns. 1 gene 2 T/F based on DEA\n\ninput2: table with two columns. 1 gene 2 gene length",
```

File: workflows/transcriptomics/goseq/README.md [hunk 16]
```diff
@@ -0,0 +1,29 @@
+# Goseq GO-KEGG Enrichment Analysis Workflow
+
+## Inputs dataset
+
+The workflow need the following inputs:
+- The DEG file:
+    - A tabular file with first column the gene symbol and second column a boolean value whether the gene is DEG or not.
```
Comment (lldelisle) [inline workflows/transcriptomics/goseq/README.md:7]:
```suggestion
    - A tabular file with first column the gene symbol and second column a boolean value whether the gene is a differentially expressed gene or not. 
```

File: workflows/transcriptomics/goseq/README.md [hunk 17]
```diff
@@ -0,0 +1,29 @@
+# Goseq GO-KEGG Enrichment Analysis Workflow
+
+## Inputs dataset
+
+The workflow need the following inputs:
+- The DEG file:
+    - A tabular file with first column the gene symbol and second column a boolean value whether the gene is DEG or not. 
+- The gene length file:
+    - A tabular file with first column the gene symbol and second column the gene length of the genes.
```
Comment (lldelisle) [inline workflows/transcriptomics/goseq/README.md:9]: Could you indicate the potential users where they can get this info? For example using a GTF and the tool "Gene length and GC content from GTF and FASTA file (Galaxy Version 0.1.2)"

File: workflows/transcriptomics/goseq/README.md [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+- The KEGG file:
+    - A tabular file with first column the Pathway ID and second column the Pathway name like:
```
Comment (lldelisle) [inline workflows/transcriptomics/goseq/README.md:11]: Same here, could you indicate where you can get this info?

File: workflows/transcriptomics/goseq/README.md [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+        -   ```
+            ID  Name
+            01100   Metabolic pathways - mmus
+            01200   Carbon metabolism - mmus 
+            ```
+ 
+## Processing
+
+- The workflow will do a simple enrichment analysis with taking into account the gene length
+- The output will be 3 files `GO table`, `Top ontology plot` and `DE genes in each category` for CC, BP, and MF ontologies and `KEGG table` and `DE genes in each KEGG Pathways`
```
Comment (lldelisle) [inline workflows/transcriptomics/goseq/README.md:21]:
```suggestion
- The output will be 3 files `GO table`, `Top ontology plot` and `DE genes in each category` for Cellular Component, Biological Processes, and Molecular Function ontologies and `KEGG table` and `DE genes in each KEGG Pathways`
```

File: workflows/transcriptomics/goseq/README.md (hunk 1 above)
Comment (lldelisle) [reply workflows/transcriptomics/goseq/README.md:25]: Would be good to also do it with the Select Gene ID format parameter.

**Review by lldelisle — COMMENTED** [review]
Thanks for this contribution. I added some comments.

**Comment by nilchia** [comment]
@mvdbeek and @lldelisle Thanks a lot for your review!

**Comment by nilchia** [comment]
I can't see the problem here. The test passes locally.

File: workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis-tests.yml [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+    Select genome to use: mm10
+    Select gene ID format: ensGene
+  outputs:
+    goseq Enrichment Table (Cellular Component):
+      asserts:
+        - that: "has_text_matching"
+          expression: "category\tover_represented_pvalue\tunder_represented_pvalue\tnumDEInCat\tnumInCat\tterm\tontology\tp_adjust_over_represented\tp_adjust_under_represented"
+        - that: "has_text_matching"
+          expression: "GO:0005737\t4.7986980823087e-08\t1\t4810\t8312\tcytoplasm\tCC\t9.31427297776118e-05\t1"
```
Comment (lldelisle) [inline workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis-tests.yml:23]:
```suggestion
          expression: "GO:0005737\t4.79[0-9]+e-08\t1\t4810\t8312\tcytoplasm\tCC\t9.31[0-9]+e-05\t1"
```

File: workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis-tests.yml [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+    goseq Top Categories plot (Biological Process):
+      asserts:
+        - that: "has_size"
+          size: 7092
+    goseq Enrichment Table (Biological Process):
+      asserts:
+        - that: "has_text_matching"
+          expression: "category\tover_represented_pvalue\tunder_represented_pvalue\tnumDEInCat\tnumInCat\tterm\tontology\tp_adjust_over_represented\tp_adjust_under_represented"
+        - that: "has_text_matching"
+          expression: "GO:0043254\t3.41984403111893e-07\t1\t239\t345\tregulation of protein-containing complex assembly\tBP\t0.00344967712964967\t1"
```
Comment (lldelisle) [inline workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis-tests.yml:53]:
```suggestion
          expression: "GO:0043254\t3.41[0-9]+e-07\t1\t239\t345\tregulation of protein-containing complex assembly\tBP\t0.0034[0-9]+\t1"
```

File: workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis-tests.yml [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+        - that: "has_text_matching"
+          expression: "category\tde_genes"
+        - that: "has_text_matching"
+          expression: "GO:0005515\tENSMUSG00000045545,ENSMUSG00000029802,ENSMUSG00000058056,ENSMUSG00000004044,"
+    goseq Enrichment Table (KEGG):
+      asserts:
+        - that: "has_text_matching"
+          expression: "category\tover_represented_pvalue\tunder_represented_pvalue\tnumDEInCat\tnumInCat\tp_adjust_over_represented\tp_adjust_under_represented\tID\tName"
+        - that: "has_text_matching"
+          expression: "04540\t0.00315476030180865\t0.998728109935746\t47\t61\t0.461560170383683\t1\t04540\tGap junction - mmus"
```
Comment (lldelisle) [inline workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis-tests.yml:89]:
```suggestion
          expression: "04540\t0.0031[0-9]+\t0.99[0-9]+\t47\t61\t0.46[0-9]+\t1\t04540\tGap junction - mmus"
```

File: workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis-tests.yml [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+        - that: "has_text_matching"
+          expression: "over_represented_pvalue"
+        - that: "has_text_matching"
+          expression: "under_represented_pvalue"
```
Comment (lldelisle) [inline workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis-tests.yml:93]: This is already in the first test with the header
```suggestion

```

File: workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis-tests.yml [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+    goseq Top Categories plot (Molecular Function):
+      asserts:
+        - that: "has_size"
+          size: 7166
+    goseq Enrichment Table (Molecular Function):
+      asserts:
+        - that: "has_text_matching"
+          expression: "category\tover_represented_pvalue\tunder_represented_pvalue\tnumDEInCat\tnumInCat\tterm\tontology\tp_adjust_over_represented\tp_adjust_under_represented"
+        - that: "has_text_matching"
+          expression: "GO:0005515\t1.40940451349193e-07\t1\t3758\t6415\tprotein binding\tMF\t0.000600265382296212\t1"
```
Comment (lldelisle) [inline workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis-tests.yml:73]:
```suggestion
          expression: "GO:0005515\t1.4094[0-9]+e-07\t1\t3758\t6415\tprotein binding\tMF\t0.0006[0-9]+\t1"
```

**Review by lldelisle — COMMENTED** [review]
It happens that the precision differs between computers for example:
4.798698082308**7**e-08 locally become 4.798698082308**69**e-08 in the CI. To overcome this, I tend to use the expression `[0-9]+`.

**Review by lldelisle — DISMISSED** [review]
Great!

**Comment by lldelisle** [comment]
@mvdbeek I let you do a final review and merge if nothing to change.

**Comment by nilchia** [comment]
@lldelisle Thanks!

File: workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis.ga [hunk 25]
```diff
@@ -0,0 +1,702 @@
...
+        {
+            "class": "Person",
+            "identifier": "https://orcid.org/0000-0003-1329-3044",
+            "name": "Amirhossein Naghsh Nilchi"
+        }
+    ],
+    "format-version": "0.1",
+    "license": "MIT",
+    "release": "0.1",
+    "name": "goseq-go-kegg-enrichment-analsis",
```
Comment (mvdbeek) [inline workflows/transcriptomics/goseq/goseq-go-kegg-enrichment-analsis.ga:15]:
```suggestion
    "name": "Goseq GO-KEGG Enrichment Analysis",
```

**Review by mvdbeek — DISMISSED** [review]
Looks good, just need a minor change to the workflow name.

**Review by mvdbeek — APPROVED** [review]

**Comment by bgruening** [comment]
Yeah! Congratulations @nilchia!

**Pull request merged**